feat: Bookkeep fixed parameters

kratsg opened the pull request:
# Description

This will help with #846 .

The goal of this PR is to:
- [x] allow for `pdf.config.fixed_pars` to provide a list of True/False of parameters which are fixed. This can be essentially stitched with the inits provided by the user, or `pdf.config.suggested_init()` to get `fixed_vals` for the optimizer API.
  - [x] tests are added
- ~[ ] update the schema (we were idiots) to change `fixed` to be an array as it needs to be defined for each parameter~
  - ~[ ] tests are added~
  - ~we can additionally allow for a single bool *and* an array of bools. The single bool could provide backwards compatibility implying all parameters for the modifier are specified as fixed.~
- ~[ ] update `src/pyhf/readxml.py` to correctly read in `ParamSetting` fixed parameters correctly~
  - ~[ ] tests are added~
- ~[ ] update `src/pyhf/writexml.py` to write out fixed parameters correctly against the new schema~
  - ~[ ] tests are added~

This PR will not propagate fixed parameters from the `pdf` through to the optimizers. That is the goal of #846.

Help requested:
- ~we need examples of HistFactory XMLs that fixed one or two parameters of a given modifier (rather than all of them).~
  - decided for now paramsetting will fix all parameters of a given systematic

Bugs identified(?)
- ~in `src/pyhf/readxml.py`, `inits` suffers a similar bug that `fixed` does, where the `inits` for a modifier with n-parameters will be a list of one value.~
  - inits (like fixed) will be for all parameters of a systematic

Brain dump:
- ~for `readxml.py`: use `re` or `parse` to parse out parameter indices in `ParamSetting.text` values. I will need an example (see "Help requested") but I assume `gamma_stat_SR0L_IStR_cuts[2]` would be written into `ParamSetting` and we need to extract the index of the parameter that such a thing is being configured for (parameter setting) -- and if there is no indexing involved -- this is assumed for all parameters in that modifier (modifier setting).~
- ~for `readxml.py`: if we parse out parameter setting, we're going to need to figure out how many parameters the associated modifier has in total (only for: staterror, shapesys, shapefactor) so we can stitch this in appropriately~
- ~perhaps to preserve backwards compatibility, we can move the complexity out of `readxml/writexml` and handle parameter parsing inside `pdf.py`. Instead of trying to go from `gamma_stat_SR0L_IStR_cuts[2]` to `gamma_stat_SR0L_IStR_cuts` (or backwards) in read/write XML... we can just write out `gamma_stat_SR0L_IStR_cuts[2]` as literally the name of the parameter configuration which our JSON is technically for, rather than modifier configuration. As such, every parameter is single-bin -- so the complexity is moved over to trying to reinflate these parameters and inject them into the corresponding modifiers.~

ReadTheDocs build: https://pyhf.readthedocs.io/en/feat-bookkeepfixedparameters

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR
```
* Add functionality into _ModelConfig to bookkeep fixed systematics (all parameters for a given systematic)
* Add fixed into tests for paramsets
* Add tests for fixed parameters
```

kratsg commented on src/pyhf/modifiers/staterror.py:
```diff
@@ -19,6 +19,7 @@ def required_parset(cls, sample_data, modifier_data):
             'is_shared': True,
             'inits': (1.0,) * len(sample_data),
             'bounds': ((1e-10, 10.0),) * len(sample_data),
+            'fixed': False,
```
For now, systematic-level application.. but we'd want to change to `(False,)*len(sample_data)` later?

kratsg commented on src/pyhf/modifiers/shapesys.py:
```diff
@@ -27,6 +27,7 @@ def required_parset(cls, sample_data, modifier_data):
             'is_shared': False,
             'inits': (1.0,) * n_parameters,
             'bounds': ((1e-10, 10.0),) * n_parameters,
+            'fixed': False,
```
For now, systematic-level application.. but we'd want to change to `(False,)*len(sample_data)` later?

kratsg commented on src/pyhf/modifiers/shapefactor.py:
```diff
@@ -19,6 +19,7 @@ def required_parset(cls, sample_data, modifier_data):
             'is_shared': True,
             'inits': (1.0,) * len(sample_data),
             'bounds': ((0.0, 10.0),) * len(sample_data),
+            'fixed': False,
```
For now, systematic-level application.. but we'd want to change to `(False,)*len(sample_data)` later?

kratsg commented on src/pyhf/parameters/paramsets.py:
```diff
@@ -2,25 +2,34 @@
         self.n_parameters = kwargs.pop('n_parameters')
         self.suggested_init = kwargs.pop('inits')
         self.suggested_bounds = kwargs.pop('bounds')
+        self.fixed = kwargs.pop('fixed')
```
NB: this requires all paramsets to be defined with `fixed` kwarg.

kratsg commented on tests/test_schema.py:
```diff
@@ -257,6 +257,34 @@ def test_parameters_duplicated():
+        ],
+        'parameters': [{'name': 'mypoi', 'inits': [1], 'fixed': True}],
+    }
+    pyhf.Model(spec, poi_name='mypoi')
```
we're checking to ensure there's no exception raised

matthewfeickert replied:
Yes, I guess I still would have assumed we'd do an assert for existence. But as I said that's not required.

lukasheinrich commented on src/pyhf/parameters/paramsets.py:
```diff
@@ -2,25 +2,34 @@
 
 
 class paramset(object):
+    __slots__ = ['n_parameters', 'suggested_init', 'suggested_bounds', 'fixed']
```
why do we need `__slots__`?

kratsg replied:
Only reason really is to save space and ensure we're not adding any other extra attributes to the paramsets. This also helps automatically document all the available properties on a paramset. In this case, it simply formalizes what we're already doing.

lukasheinrich commented on src/pyhf/parameters/paramsets.py:
```diff
@@ -2,25 +2,34 @@
 
 
 class unconstrained(paramset):
+    __slots__ = ['constrained']
```
why do we need `__slots__`?

kratsg replied:
Only reason really is to save space and ensure we're not adding any other extra attributes to the paramsets. This also helps automatically document all the available properties on a paramset. In this case, it simply formalizes what we're already doing.

lukasheinrich replied:
save space?

kratsg replied:
https://docs.python.org/3/reference/datamodel.html#slots

lukasheinrich commented on src/pyhf/parameters/utils.py:
```diff
@@ -27,6 +28,7 @@ def reduce_paramsets_requirements(paramsets_requirements, paramsets_user_configs
         combined_paramset = {}
         for k in paramset_keys:
             for paramset_requirement in paramset_requirements:
+                # undefined: the modifier does not support configuring that property
```
what does this comment refer to?

kratsg replied:
just a reminder of what 'undefined' meant in this context... I forgot. It'll help when going through this code again in the future.

lukasheinrich replied:
ok

lukasheinrich commented on src/pyhf/pdf.py:
```diff
@@ -216,6 +217,16 @@ def _nominal_and_modifiers_from_spec(config, spec):
+        'auxdata',
+        'auxdata_order',
+        'modifier_settings',
+    ]
```
do we want to keep `slots` here?

matthewfeickert replied:
@kratsg

lukasheinrich commented:
LGTM

alexander-held commented:
Hi, just gave this a try. I think `pdf.config.fixed_pars` in the description should mean `pdf.config.suggested_fixed()`. I wonder whether "suggested" could be misleading, but I guess it depends on whether one views the `measurements` part of the workspace as a suggestion for how to perform a measurement, or one views the whole workspace as an instruction for doing one specific fit (I personally perceive it more like the second option).

matthewfeickert commented:
> Hi, just gave this a try. I think `pdf.config.fixed_pars` in the description should mean `pdf.config.suggested_fixed()`. I wonder whether "suggested" could be misleading, but I guess it depends on whether one views the `measurements` part of the workspace as a suggestion for how to perform a measurement, or one views the whole workspace as an instruction for doing one specific fit (I personally perceive it more like the second option).

@alexander-held Can you open this up as an Issue for us to discuss there (also hopefully making it easier for others to find for future reference)?

kratsg commented:
> Hi, just gave this a try. I think `pdf.config.fixed_pars` in the description should mean `pdf.config.suggested_fixed()`

Fixed.

> I wonder whether "suggested" could be misleading, but I guess it depends on whether one views the `measurements` part of the workspace as a suggestion for how to perform a measurement, or one views the whole workspace as an instruction for doing one specific fit (I personally perceive it more like the second option).

Part of this is to separate out a model definition from a measurement definition. The measurement definitions come into play with `suggested_xyz` accessors -- as you probably saw (inits, bounds, fixed).

alexander-held commented:
I'll open an issue as suggested by @matthewfeickert, but in the end it's matter of taste what to call it I think.
